### ISSUE TITLE
Sampling from discrete distribution and prior sensitivities

### ISSUE THREAD

**Issue by dmi3kno** [greta-dev/greta#191]
Hi and thank you for the awesome package!

Following are really really silly questions related to non-standard use of the package. Please, feel free to close this issue as "off-topic".

I am thinking of using `greta` for decision analysis. Lets say we have uncertain price of an asset and discrete possible acquired stake 0, 50% or 100%. I want to simulate the deal value (The distribution itself does not mean much).

```r
library(greta)

# yay, I can define some sensitivities
price = normal(c(2,3),c(1,3))

probs <- c(0.4, 0.35, 0.25)
cps <- cumsum(c(0, probs[-3]))

# Vector of possible acquired stakes
# p <- c(0, 0.5, 1)
# with probabilities probs
i = sum(uniform(0,1)>= cps)
s =(i-1)/2

# deal_value = price * stake
vn = price*s

m <- model(price, s, vn)

draws <- mcmc(m, n_samples = 1000)

## ggplot2 code for plotting ecdf
```
![rplot01](https://user-images.githubusercontent.com/13419011/41927901-d9517fe2-7973-11e8-8d1e-cbb2e162a4a6.png)

It seems like I can simulate several priors and (with some rearranging) plot and compare them. Not sure how this would go if I wanted to run alternative prior (different distribution, not just different parameters). What do you think my approach should be?

I tried very hard to sample from discrete distribution. My object `i` contains indices to the vector of discrete values `p`, but I can't use subsetting to "look it up". I am doing some math in `s`, but that is just a numerical coincidence. What is the best way to use simulated values for subsetting? What if I define my `p` as `greta.array`, would I be able to subset it with simulated values in `i`?

Finally, entirely stupid question regarding channels: can I use them for sensitivities (i.e. alternative priors or other parameters for simulation). 

I have not gotten into likelihoods yet, but I would eventually like to be able to update discrete priors with discrete likelihoods and sample from the resulting posterior.

**Comment by goldingn** [comment]
Heya, sorry for being slow to reply. I've been trying to find time to think about these issues more deeply, since I'm not familiar with these models.

You've hit on a limitation of greta here - because the current version only allows HMC, discrete variables and distributions aren't allowed (you can't use HMC on a model with discrete variables, because you can't get a gradient for the discrete parameters). So I'm not sure you're going to be able to use greta for this problem, until I implement discrete variables and samplers (#104) 

What you've got coded up now is valid, since you're sampling continuous variables and then doing things to them - though using HMC is of course an inefficient way of generating standard uniform random variables. If you were to use anything after `i` in a likelihood, you wouldn't be able to use HMC because the derivatives of the model joint density wouldn't be defined.

There's an open issue (#111) that I really need to resolve, which means that greta probably wouldn't throw an error if you added a likelihood term in there, but may well give the wrong result. greta will error on defining the model if you explicitly use a discrete unfixed variable (e.g. `i <- sum(bernoulli(cps))`), but it can't currently detect use of discrete operations (`<=`) in the model, and in some of these situations, tensorflow can compute a gradient, but it's not the one we want :|

When discrete random variables are available, I'll make it so that you can then use them for subsetting other greta arrays.

I'm sorry, but I don't quite get what you're asking with these two Qs:

> how this would go if I wanted to run alternative prior (different distribution, not just different parameters). What do you think my approach should be?

Do you mean how can you hack together discrete random variables following other distributions? Again, best to wait for discrete variables to be implemented I think.

> channels: can I use them for sensitivities (i.e. alternative priors or other parameters for simulation).

I don't know what 'channels' are in this context, nor what you mean by sensitivities. Could you point me to a definition so I can get my head around these models?

**Comment by bearloga** [comment]
Out of curiosity, would it possible to have discrete latent variables without marginalization if the user could sample via [tfp.mcmc.MetropolisHastings](https://www.tensorflow.org/probability/api_docs/python/tfp/mcmc/MetropolisHastings) instead of HMC?

**Comment by goldingn** [comment]
Yes, you could certainly hack your way to discrete RVs and use MH. I'm not entirely sure the sampler will converge on the correct distribution (I think it will, but haven't thought hard about that) and in any case I'd be surprised if it tuned and sampled well. I'd be happy to be proven wrong though!

Explicit handling of discrete RVs is high on the to do list (should have a prototype early next year), and that will include MCMC proposals that are designed for discrete RVs, and the ability to interleave discrete and continuous sampling in the MCMC so you can still use the magic of HMC for continuous RVs.

**Comment by goldingn** [comment]
In case you haven't seen it, you can do MH now in greta with `rwmh()`.
